base/src/com/google/idea/blaze/base/qsync/ProjectUpdater.java

```diff
@@ -135,90 +142,98 @@ private void updateProjectModel(ProjectProto.Project spec, Context<?> context) {
+                      roots.addContentEntry(
+                          UrlUtil.pathToUrl(projectPathResolver.resolve(projectPath).toString()));
+                  for (ProjectProto.SourceFolder sfSpec : ceSpec.getSourcesList()) {
+                    ProjectPath sourceFolderProjectPath = ProjectPath.create(sfSpec.getProjectPath());
 
-            for (String lib : moduleSpec.getLibraryNameList()) {
-              Library library = libMap.get(lib);
-              if (library == null) {
-                throw new IllegalStateException(
-                    "Module refers to library " + lib + " not present in the project spec");
-              }
-              LibraryOrderEntry entry = roots.addLibraryEntry(library);
-              // TODO should this stuff be specified by the Module proto too?
-              entry.setScope(DependencyScope.COMPILE);
-              entry.setExported(false);
-            }
+                    JavaSourceRootProperties properties =
+                        JpsJavaExtensionService.getInstance()
+                            .createSourceRootProperties(
+                                sfSpec.getPackagePrefix(), sfSpec.getIsGenerated());
+                    JavaSourceRootType rootType =
+                        sfSpec.getIsTest() ? JavaSourceRootType.TEST_SOURCE : JavaSourceRootType.SOURCE;
+                    String url =
+                        UrlUtil.pathToUrl(
+                            projectPathResolver.resolve(sourceFolderProjectPath).toString(),
+                            sourceFolderProjectPath.innerJarPath());
+                    SourceFolder unused = contentEntry.addSourceFolder(url, rootType, properties);
+                  }
+                  for (String exclude : ceSpec.getExcludesList()) {
+                    contentEntry.addExcludeFolder(
+                        UrlUtil.pathToIdeaDirectoryUrl(workspaceRoot.absolutePathFor(exclude)));
+                  }
+                }
 
-            WorkspaceLanguageSettings workspaceLanguageSettings =
-                LanguageSupport.createWorkspaceLanguageSettings(projectViewSet);
+                for (String lib : moduleSpec.getLibraryNameList()) {
+                  Library library = libMap.get(lib);
+                  if (library == null) {
+                    throw new IllegalStateException(
+                        "Module refers to library " + lib + " not present in the project spec");
+                  }
+                  LibraryOrderEntry entry = roots.addLibraryEntry(library);
+                  // TODO should this stuff be specified by the Module proto too?
+                  entry.setScope(DependencyScope.COMPILE);
+                  entry.setExported(false);
+                }
+                return new AbstractMap.SimpleImmutableEntry<>(module, moduleSpec);
+              }).toList();
+          return new AbstractMap.SimpleImmutableEntry<>(models, modules);
+        },
+        readValue -> {
```

Comment:
We're in the middle of a method call which spans more than 100 lines. This whole class is in need of refactoring.